src/librustc_lint/builtin.rs

```diff
@@ -1927,8 +1932,22 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for InvalidValue {
                             return Some(InitKind::Zeroed);
                         }
                     }
-                    // FIXME: Also detect `MaybeUninit::zeroed().assume_init()` and
-                    // `MaybeUninit::uninit().assume_init()`.
+                }
+            } else if let hir::ExprKind::MethodCall(ref path, _, ref args) = expr.kind {
+                // Find problematic calls to `MaybeUninit::assume_init`.
+                if path.ident.name == sym::assume_init {
```

Comment:
This one seems odd to me. I am comparing just the name of the method -- but what if there is a trait or another type or so that also had this method? This looks like I am working pre-resolution, which seems fragile. Isn't there any way to get the full path to the method that will actually be called here?

I looked at clippy but couldn't find anything less fragile there either. Cc @oli-obk @Manishearth

Reply:
Can you use diagnostics items instead? https://github.com/rust-lang/rust/pull/60966

Reply:
Hm, I have an `Option<HirId>`; I guess I could unwrap that and then call `owner_def_id` to get the `DefId` that `is_diagnostic_item` needs.

Reply:
Nope, that does not work. The lint just never fires. Looks like the `HirId` I get there is not pointing to the method being called -- that kind of makes sense if the information we have here is pre-method-resolution. (Probably it's the `HirId` of the method call itself.)

For `Call`, I can call
```rust
cx.tables.qpath_res(qpath, path_expr.hir_id).opt_def_id()?
```
to get the `def_id` of the callee. Isn't there something similar for `MethodCall`?

Reply:
Hm, maybe I need to construct a `QPath::TypeRelative`? I have a `PathSegment`, but I'd still also need a `Ty`.

Reply:
Ah, I found a way:
```rust
cx.tables.type_dependent_def_id(expr.hir_id)?
```
Here, `expr` is the `MethodCall` itself.